migrations/20231128_jellyfish_migration/utils/data_migration.go

```diff
@@ -0,0 +1,386 @@
+		Rollback:            true,
+		RollbackSectionName: "_rollbackMigration",
+		DryRun:              true,
+		StopOnErr:           true,
+		QueryBatchSize:      50,
+		QueryBatchLimit:     100,
+	}
+	if dryRun != nil {
+		settings.DryRun = *dryRun
+	}
+	if stopOnErr != nil {
+		settings.StopOnErr = *stopOnErr
+	}
+	if rollback != nil {
+		settings.Rollback = *rollback
+	}
+	if rollbackSectionName != nil {
+		settings.RollbackSectionName = *rollbackSectionName
+	}
+	if writeToDisk != nil {
+		settings.writeToDisk = *writeToDisk
+	}
+	if queryBatch != nil {
+		settings.QueryBatchSize = *queryBatch
+	}
+	if queryLimit != nil {
+		settings.QueryBatchLimit = *queryLimit
+	}
+	if capacity != nil && *capacity > 0 {
+		settings.capacity = capacity
+		log.Printf("capped at %d items", *settings.capacity)
+	}
+	return settings
+}
+
+type DataMigration struct {
+	ctx                  context.Context
+	dataC                *mongo.Collection
+	settings             *Settings
+	updates              []mongo.WriteModel
+	groupedDiffs         map[string][]UpdateData
+	groupedErrors        groupedErrors
+	errorsCount          int
+	updatedCount         int
+	fetchedCount         int
+	lastUpdatedID        string
+	startedAt            time.Time
+	mongoInstanceChecker MongoInstanceCheck
+}
+
+type DataMigrationQueryFn func(m *DataMigration) bool
+type DataMigrationUpdateFn func(m *DataMigration) (int, error)
+
+type groupedErrors map[string][]ErrorData
+
+func NewMigration(ctx context.Context, settings *Settings, checker MongoInstanceCheck, dataC *mongo.Collection, lastID *string) (*DataMigration, error) {
+	var err error
+	if settings == nil {
+		err = errors.Join(err, errors.New("missing required settings"))
+	}
+	if checker == nil {
+		err = errors.Join(err, errors.New("missing required mongo checker"))
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	m := &DataMigration{
+		ctx:                  ctx,
+		dataC:                dataC,
+		mongoInstanceChecker: checker,
+		settings:             settings,
+		updates:              []mongo.WriteModel{},
+		groupedErrors:        groupedErrors{},
+		groupedDiffs:         map[string][]UpdateData{},
+		errorsCount:          0,
+		updatedCount:         0,
+		fetchedCount:         0,
+		startedAt:            time.Now(),
+	}
+	if lastID != nil {
+		m.lastUpdatedID = *lastID
+	}
+	return m, nil
+}
+
+func (m *DataMigration) Initialize() error {
+	if err := m.mongoInstanceChecker.CheckFreeSpace(m.ctx, m.dataC); err != nil {
+		return err
+	}
+	if err := m.mongoInstanceChecker.SetWriteBatchSize(m.ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *DataMigration) GetCtx() context.Context {
+	return m.ctx
+}
+
+func (m *DataMigration) GetSettings() Settings {
+	m.settings.WriteBatchSize = m.mongoInstanceChecker.GetWriteBatchSize()
+	return *m.settings
+}
+
+func (m *DataMigration) GetDataCollection() *mongo.Collection {
+	return m.dataC
+}
+
+func (m *DataMigration) Execute(
+	queryFn DataMigrationQueryFn,
```

Comment:
Using functions a bit harder to understand, but okay, I suppose. This is for testing right? Next time consider interfaces.